### PR TITLE
Restrict user defined DS class

### PR DESCRIPTION
**Pull request by junmendoza**
### Purpose

1. This PR restricts the use of DS defined classes only if it uses the "InternalClass" attribute

2. Currently, only the classes in FunctionObject.ds contain this attribute

3. Category DSDefinedClass is already added for all tests that use DS defined classes

### FYI

On a test run in SelfServe CI where all DSDefinedClass tests are ignored, there are no other regressions in DS ProtoTest and Dynamo tests

### Impt note

Once this is acceptable, a new PR must be created to add an Ignore flag to all tests with the DSDefinedClass category. That must be merged first before this PR 

### Reviewers

@lukechurch 
@ke-yu

### CODE REVIEW

File: src/Engine/ProtoCore/ExtendedLibraries/FunctionObject.ds [hunk 1]
```diff
@@ -1,3 +1,5 @@
+
+[InternalClass]
```
Comment (sharadkjaiswal) [inline src/Engine/ProtoCore/ExtendedLibraries/FunctionObject.ds]: Great, I didn't know that this syntax was supported. Can we have a similar attribute on function definition, to hide the internal methods from the library view. Currently, any method starting with underscore (_) is treated internal.

**Comment by lukechurch** [comment]
@junmendoza This looks ok, but I'm concerned to do the bulk of the tedious work fixing up the classes to use FFI backed ones before turning off the tests. 

Could we start that exercise first? Do you know how many tests it is?

**Comment by junmendoza** [comment]
@lukechurch Yes the task is planned this sprint: http://adsk-oss.myjetbrains.com/youtrack/issue/MAGN-8040

The strategy is to do the following:
1. Identify all tests that share a common pattern of using DS defined classes and create an FFI class what will be imported from these tests
2. Identify test that import a ds file where the DS file includes a DS defiend class and replace that with an FFI implementation

File: src/Engine/ProtoCore/ExtendedLibraries/FunctionObject.ds (hunk 1 above)
Comment (junmendoza) [inline src/Engine/ProtoCore/ExtendedLibraries/FunctionObject.ds]: Yes this feature supports both DS classes and functions. This is possible as a separate task as it also involved creating an IgnoreFunctionAttribute. There should be minimal cost to performance as it only occurs at compile time - when the DS code is compiled

**Comment by junmendoza** [comment]
FYI: This will require sifting through 3260 DS testcases. I dont know how long this will take but perhaps we can  limit the minimum tests we need to fix prior to merging in this PR?

File: src/Engine/ProtoAssociative/CodeGen.cs [hunk 2]
```diff
@@ -5382,11 +5382,42 @@ private void EmitMemberVariables(ClassDeclNode classDecl, GraphNode graphNode =
...
+            {
+                return true;
+            }
+
+            // Check the class attributes
+            if (classDecl.Attributes != null)
+            {
+                // If at least one attribute is internal then the class is allowed
+                List<AttributeEntry> attributesList = PopulateAttributes(classDecl.Attributes);
+                if (attributesList.Where(a => a.IsInternalClassAttribute()).Count() > 0)
```
Comment (ke-yu) [inline src/Engine/ProtoAssociative/CodeGen.cs]: Do we need to PopulateAttributes? Can we write
```
classDecl.Attributes.Any(condition)
```
?

**Comment by lukechurch** [comment]
@junmendoza Lets port the first 80% and see where we get to.

**Comment by ke-yu** [comment]
hi @lukechurch , as we already port all classes, i'm going to verify if there is regression in this branch or not and merge it.

**Comment by ke-yu** [comment]
Move to the other PR: https://github.com/DynamoDS/Dynamo/pull/5701

**Pull request closed without merging**